# Notebook 2 - Analysis of a text file

The purpose of this section is to do basic text analysis with a practical purpose.

In this notebook you will:
* Find the number of occurrences of the word 'and'
* Find the number of occurrences of any word
* Construct a dictionary of all words that have appeared in the text
* Attempt to clean up the text, leaving only words

For this we will be using a text called [Humanistic Nursing by Josephine G. Paterson and Loretta T. Zderad](http://www.gutenberg.org/ebooks/25020). You should already have the file uploaded to your workspace. If that is not the case, you can download the `.txt` file from the link provided.

Estimated duration: 1h

## Dealing with files
First, we will need to learn how to operate with files. Python offers many versatile ways and here we will touch on the basics needed for the exercise.

We can open a file simply by using the `open()` function and supplying it with the path of the desired file. 

Python will start looking in the same directory (folder) as the current notebook by default. The current directory has the file path `.`, and a file in a subdirectory of the current one would have the file path `./subdirectory/file_name.type`. If the current directory lives in another directory, its parent directory is accessed by the filepath `..`.

By default the function opens files in read-only mode `"r"` and reads the text by lines.

PS. No need to read the full text.

In [1]:
filename = "./data/humanistic_nursing.txt"
file = open(filename)

# print the file line by line
for line in file:
    print(line)


The Project Gutenberg eBook, Humanistic Nursing, by Josephine Paterson, and Loretta Zderad





This eBook is for the use of anyone anywhere at no cost and with

almost no restrictions whatsoever.  You may copy it, give it away or

re-use it under the terms of the Project Gutenberg License included

with this eBook or online at www.gutenberg.org



** This is a COPYRIGHTED Project Gutenberg eBook, Details Below **

**     Please follow the copyright guidelines in this file.     **









Title: Humanistic Nursing





Authors: Josephine Paterson and Loretta Zderad







Release Date: April 8, 2008  [eBook #25020]



Language: English





***START OF THE PROJECT GUTENBERG EBOOK HUMANISTIC NURSING***





Copyright (C) 2007 by Josephine Paterson and Loretta Zderad.







Humanistic Nursing



(Meta-theoretical Essays on Practice)



by Josephine Paterson and Loretta Zderad



Copyright (C) 2007 by Josephine Paterson and sLoretta Zderad all rights

reserved except as follows.  This 


presence, as well as in nursing care activities. The nurse feels in

harmony with the rhythm of the dialogue and, sensing the timing of its

flow, she paces her call and response to patient's ability to call and

respond in that moment. So, as a nurse, you may find yourself almost

unconsciously or intuitively waiting, holding back, anticipating, urging

the patient. This kind of synchronization or timing is intersubjective

for the clues or reasons for encouraging or waiting are not found solely

in the patient's behavior nor only in the nurse's knowledge or

experience. "Good" or "right" timing somehow involves the "between." It

implies that nurse and patient share not only clock time but private,

lived time.





Space



By exploring the dialogue of nursing as it is lived in the real world

the factor of space becomes apparent. Here again the dialogue is

influenced by space as it is measured and space as it is experienced by

nurse and patient. When thinking of health care faci



About our inadequacies of expression, many things are, are true,

"all-at-once." The law of contradiction does not apply

in-the-lived-experienced-world. We each view the world through our

unique histories. Wisdom is many sided truth. Wisdom cannot be expressed

"all-at-once." Truths can be stated only in sequence or metaphorically.

If I were supercritical of my human limitations to express "all-at-once"

wisdom, I would say nothing. Jung points up the dangers of this, he

says:



     "I must prevent my critical powers from destroying my

     creativeness. I know well enough that every word I utter

     carries with it something of myself--of my special and unique

     self with its particular history and its particular world."[7]



Each nurse's uniqueness dictates then a responsibility to share her

particular knowing with fellow struggling human beings. Only through

each describing can there be correction and complementary synthesis to

movement beyond.



The nurse's worl


dancers see, hear, feel in the nursing dialogue?





Therapeutics



There is a third way in which humanistic nursing and art are related.

For many years, the arts have been used in nursing for their therapeutic

effects, especially with psychiatric, geriatric, and pediatric patients.

The nurse and a patient or a group of patients participate in an

artistic experience together. These may be passive activities, such as,

attending a concert or play or visiting {89} an art exhibit; or they

may be active ones in which nurse and patients are involved in artistic

expression or creation.



Music, poetry, painting, drama, and dance have been used effectively in

various nursing situations. For instance, Christoffers, a nurse and

dancer, emphasizes the importance of body language as communication and

supports her view with clinical evidence. She urges nurses to become

"physically literate--to develop an understanding and appreciation of

the part played by body language in human rel

replied, "No, well I had a social worker when I was a little girl." I

tried at this point to transfer feelings of the past to the present.

"Oh, for how long? What was she like?" "I don't remember," and Irene

closed her eyes. In a few minutes Irene requested that I set her hair.

She is capable of doing this herself. I set her hair, but discussed the

question of what she was really asking for. I believe she was asking for

concrete attention to test my ability to care for her. I was trying to

say, concretely, by setting her hair, that people could care about her.





     9. I encouraged patients' expression to come to understand

     better their behavioral messages to enable me to respond

     overtly as appropriately and therapeutically as possible.



_Example_



The previous time I was at the hospital Alice had not come to the unit.

I was told that she felt too depressed to come down. I went to see her.

She had looked surprised and impressed by my visit. She talked on at

Let us skip reading the text for the time being.

In the example about we opened the file in read-only mode. What if we want to write to a file? Then we must open it in write-only mode(`"w"`):
```python
file = open(filename, "w")
```

but be careful with this, as this will overwrite any file with that filename. In our case if we use
```python
filename = "./data/humanistic_nursing.txt"
file = open(filename, "w")
```
we would see that the file previous file will have been **deleted**. An alternative of that is to use the `"x"` write-only mode which again creates a new file for writing but fails if that file already exists.

Here is a table of all available Python file modes:

| Mode | Description |
|----|:--|
| r  | Read-only mode |
| w  | Write-only mode;<br>creates a new file (erasing the data for any file with the same name) |
| x  | Write-only mode;<br>creates a new file, but fails if the file path already exists |
| a  | Apend to existing file; creates the file if it does not already exist |
| r+  | Read and write |
| b | Add to mode for binary files (i.e., `rb` or `wb`) |
| t | Text mode for files (automatically decoding bytes to Unicode). <br> This is the default if not specified. <br> Add `t` to other modes to use this (i.e., `rt` or `xt` ) |


Once a file is opened we can read it in a couple of ways:
- `read()` - returns a certain number of characters from the file.

In [28]:
filename = "./data/humanistic_nursing.txt"
file = open(filename)

print(file.read(10))

The Proje


`read()` advances the file handle's position by the number of bytes read.
- `tell()` - gives you the current position in a file

In [3]:
print(file.tell())

12


In [4]:
# if we read from the file again, we will advance the handle
print(file.read(5))
print(file.tell())

ct Gu
17


- `seek()` - changes the position of the file's handle

In [5]:
file.seek(3)
file.tell()

3

Finally, after you are done with a file is important to close it. If you don't then the file would not be accessible elsewhere.

In [6]:
file.read()

'The Project Gutenberg eBook, Humanistic Nursing, by Josephine Paterson, and Loretta Zderad\n\n\nThis eBook is for the use of anyone anywhere at no cost and with\nalmost no restrictions whatsoever.  You may copy it, give it away or\nre-use it under the terms of the Project Gutenberg License included\nwith this eBook or online at www.gutenberg.org\n\n** This is a COPYRIGHTED Project Gutenberg eBook, Details Below **\n**     Please follow the copyright guidelines in this file.     **\n\n\n\n\nTitle: Humanistic Nursing\n\n\nAuthors: Josephine Paterson and Loretta Zderad\n\n\n\nRelease Date: April 8, 2008  [eBook #25020]\n\nLanguage: English\n\n\n***START OF THE PROJECT GUTENBERG EBOOK HUMANISTIC NURSING***\n\n\nCopyright (C) 2007 by Josephine Paterson and Loretta Zderad.\n\n\n\nHumanistic Nursing\n\n(Meta-theoretical Essays on Practice)\n\nby Josephine Paterson and Loretta Zderad\n\nCopyright (C) 2007 by Josephine Paterson and sLoretta Zderad all rights\nreserved except as follows.  This 

Here is a list of all methods available to files

| Method | Description |
|----|:--|
| read([size])  | Return data from file as a string, with optional size argument <br>indicating the number of bytes to read. |
| readlines([size]) | Return list of lines in the file, with optional `size` argument. |
| write(str) | write passed string to file. |
| close() | Close the handle. |
| flush() | Flush the internal I/O buffer to disk. |
| seek(pos) | Move to indicated file position(integer) |
| tell() | Return current file position as integer|
| closed | True if the file is closed |

## Counting "and"
Let's see how we can count the occurances of `and` in the file. Run the cell below.

_Note: Remember that you can deconstruct a line into individual words using the `split()` method - ie. `line.split()`._

In [ ]:
def countAnd(file_path):
    counter = 0
    file = open(file_path)
    
    for line in file:
        for word in line.split():
            if word == "and":
                counter += 1
                
    return counter

In [ ]:
# Verify your function
countAnd("./data/humanistic_nursing.txt")

Hopefully you will see 1922 for `"and"`. What happens if you search for `"And"` instead?

## Exercise 1 : Count any word
Based on the function above, now write your own function which counts the occurances of any word. For example:
```python
countAny(filename, "medicine")
```
will return the occurences of the word "medicine" in the file filename.

In [15]:
def countAny(file_path, des_word):
    counter = 0 
    file = open(file_path)
    
    for line in file: 
        for word in line.split():
            if word == des_word:
                counter += 1
    return counter

In [16]:
# Verify your function
countAny("./data/humanistic_nursing.txt", "patient")

125

Patient should occur 125 times in the text. You can also try using the function for other words!

## Exercise 2: Count multiple words
Before this exercise you should be familiar with Python dictionaries. If you're not, please see [here](https://docs.python.org/3/tutorial/datastructures.html#dictionaries).

Write a function which takes a file path and a list of words, and returns a dictionary mapping each word to its frequency in the given file.

Intuitively, we can first fill in the dictionary keys with the words in our list. Afterwards we can count the occurrences of each word and and fill in the appropriate dictionary value.

*Hint: Can we use `countAny()` for this?*

In [20]:
def countAll(file_path, words):
    counter = {}
    for word in words:
        counter[word] = countAny(filename, word)                
    return counter
    


In [21]:
# Verify your function
countAll("./data/humanistic_nursing.txt", ["patient", "and", "the"])

{'patient': 125, 'and': 1922, 'the': 2604}

You should expect `{'patient': 125, 'and': 1922, 'the': 2604}`

## Exercise 3: Cleaning up 
Unless you have already accounted for it, your counter was thrown off by some words. Consider this:

In [22]:
print(countAny(filename, "work"))
print(countAny(filename, "work."))

58
6


Although the word is the same, you have different values for the 2 keys.

In order to fix this we have to clean up the words before they are counted.

There are multiple ways to do this. You can get some ideas from the [String methods page](https://docs.python.org/3/library/stdtypes.html#string-methods)

Some things to consider:
- Do you account for upper and lowercase versions of the same word?
- Can you clean up the text before splitting it into words?
- Consider writing a seperate function for cleaning the text. This will allow you to test how effective your cleaning is.
- Are you accounting for the copyright text at the beginning and ending of the file?
- Are there any cases where cleaning up the text could give ambigious results?

Now write a function that opens up the text, cleans all of the words and returns a big long list of words:

In [25]:
def cleanText(filename):
    word_list = []
    
    with open(filename) as file:
        for line in file:
            for word in line.split():
                word_list.append(word.strip())
    
    return word_list

In [26]:
# Verify your function
cleanText(filename)

['\ufeffThe',
 'Project',
 'Gutenberg',
 'eBook,',
 'Humanistic',
 'Nursing,',
 'by',
 'Josephine',
 'Paterson,',
 'and',
 'Loretta',
 'Zderad',
 'This',
 'eBook',
 'is',
 'for',
 'the',
 'use',
 'of',
 'anyone',
 'anywhere',
 'at',
 'no',
 'cost',
 'and',
 'with',
 'almost',
 'no',
 'restrictions',
 'whatsoever.',
 'You',
 'may',
 'copy',
 'it,',
 'give',
 'it',
 'away',
 'or',
 're-use',
 'it',
 'under',
 'the',
 'terms',
 'of',
 'the',
 'Project',
 'Gutenberg',
 'License',
 'included',
 'with',
 'this',
 'eBook',
 'or',
 'online',
 'at',
 'www.gutenberg.org',
 '**',
 'This',
 'is',
 'a',
 'COPYRIGHTED',
 'Project',
 'Gutenberg',
 'eBook,',
 'Details',
 'Below',
 '**',
 '**',
 'Please',
 'follow',
 'the',
 'copyright',
 'guidelines',
 'in',
 'this',
 'file.',
 '**',
 'Title:',
 'Humanistic',
 'Nursing',
 'Authors:',
 'Josephine',
 'Paterson',
 'and',
 'Loretta',
 'Zderad',
 'Release',
 'Date:',
 'April',
 '8,',
 '2008',
 '[eBook',
 '#25020]',
 'Language:',
 'English',
 '***START',
 '

## Exercise 4: The longest word
Create a function to find and return the longest word. You are intended to reuse functions you previously made!

In [37]:
def longestWord(file_path):
    words = cleanText(filename)
    words_len = []
    for word in words: 
        words_len.append(len(word))
    # [ WRITE YOUR CODE HERE ]
    return max(words_len)
    
longestWord(filename) 


43

## Plotting with matplotlib

Let's now see how you can plot in Python, using a library called `matplotlib`. It is fairly intuitive and we can start by plotting a histogram of all word occurances in the text. For that we will use the functions you made previously.

To make computation faster we can find the occurances of just the first 100 words.

In [38]:
import matplotlib.pyplot as plt

# clean up all the text and extract a dictionary of
# word occurances
all_words = cleanText(filename)
word_dict = countAll(filename, all_words[:10])

# get a list of the words we counted and their frequencies
word_list = list(word_dict.keys())
word_freq_list = list(word_dict.values())
                 
# create the plot
plt.bar(word_list, word_freq_list)
plt.xticks(rotation=45) # rotate the x axis labels
plt.show()

<Figure size 640x480 with 1 Axes>